File: packages/plugin-obsessiondb/src/auth/api-client.ts
interface DeviceCodeResponse {
  device_code: string
  user_code: string
  verification_uri: string
  verification_uri_complete: string
  expires_in: number
  interval: number
}

interface SessionResponse {
  user: {
    id: string
    name: string
    email: string
  }
  session?: {
    activeOrganizationId?: string | null
  }
}

interface VerifiedUser {
  id: string
  email: string
  name?: string
}

export interface OtpVerifyResult {
  token: string
  user: VerifiedUser
}

/** Thrown when the send-OTP endpoint is rate-limited (HTTP 429). */
export class OtpRateLimitError extends Error {
  constructor() {
    super('Too many code requests. Please wait a minute and try again.')
    this.name = 'OtpRateLimitError'
  }
}

const CLIENT_ID = 'chkit-cli'

function userAgent(): string {
  // Avoid importing package.json — use a hardcoded prefix; version is non-critical here
  return 'chkit-cli'
}

export async function requestDeviceCode(baseUrl: string): Promise<DeviceCodeResponse> {
  const res = await fetch(`${baseUrl}/api/auth/device/code`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': userAgent(),
    },
    body: JSON.stringify({ client_id: CLIENT_ID }),
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(`Failed to request device code: ${res.status} ${text}`)
  }
  return (await res.json()) as DeviceCodeResponse
}

type TokenPollError = 'authorization_pending' | 'slow_down' | 'access_denied' | 'expired_token'

interface TokenPollResponse {
  access_token?: string
  error?: TokenPollError
}

export async function pollDeviceToken(
  baseUrl: string,
  deviceCode: string,
  interval: number,
  expiresIn: number
): Promise<string> {
  const deadline = Date.now() + expiresIn * 1000
  let pollInterval = interval * 1000

  while (Date.now() < deadline) {
    await new Promise((r) => setTimeout(r, pollInterval))

    const res = await fetch(`${baseUrl}/api/auth/device/token`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': userAgent(),
      },
      body: JSON.stringify({ client_id: CLIENT_ID, device_code: deviceCode, grant_type: 'urn:ietf:params:oauth:grant-type:device_code' }),
    })

    const body = (await res.json()) as TokenPollResponse

    if (!body.access_token && !body.error) {
      throw new Error(`Token poll failed: ${res.status} ${JSON.stringify(body)}`)
    }

    if (body.access_token) return body.access_token

    switch (body.error) {
      case 'authorization_pending':
        continue
      case 'slow_down':
        pollInterval += 5000
        continue
      case 'access_denied':
        throw new Error('Authorization denied by user.')
      case 'expired_token':
        throw new Error('Device code expired. Please try again.')
      default:
        throw new Error(`Unexpected token poll response: ${JSON.stringify(body)}`)
    }
  }

  throw new Error('Device code expired. Please try again.')
}

export async function getSession(baseUrl: string, token: string): Promise<SessionResponse> {
  const res = await fetch(`${baseUrl}/api/auth/get-session`, {
    headers: {
      Authorization: `Bearer ${token}`,
      'User-Agent': userAgent(),
    },
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(`Failed to get session: ${res.status} ${text}`)
  }
  return (await res.json()) as SessionResponse
}

/**
 * Passwordless signup/login — step 1: request a one-time code by email.
 * `type: "sign-in"` covers both new and existing users (unknown emails create a verified user).
 */
export async function sendVerificationOtp(baseUrl: string, email: string): Promise<void> {
  const res = await fetch(`${baseUrl}/api/auth/email-otp/send-verification-otp`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': userAgent(),
    },
    body: JSON.stringify({ email, type: 'sign-in' }),
  })
  if (res.status === 429) throw new OtpRateLimitError()
  if (!res.ok) {
    const text = await res.text()
    throw new Error(`Failed to send verification code: ${res.status} ${text}`)
  }
}

/**
 * Passwordless signup/login — step 2: verify the code.
 * The bearer credential is returned in the `set-auth-token` response header (not the JSON body).
 */
export async function verifyOtp(baseUrl: string, email: string, otp: string): Promise<OtpVerifyResult> {
  const res = await fetch(`${baseUrl}/api/auth/sign-in/email-otp`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': userAgent(),
    },
    body: JSON.stringify({ email, otp }),
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(`Failed to verify code: ${res.status} ${text}`)
  }
  const token = res.headers.get('set-auth-token')
  if (!token) {
    throw new Error('Verification succeeded but no auth token was returned by the server.')
  }
  const body = (await res.json()) as { user: VerifiedUser }
  return { token, user: body.user }
}

export async function createOrganization(
  baseUrl: string,
  token: string,
  input: { name: string; slug: string },
): Promise<{ id: string }> {
  const res = await fetch(`${baseUrl}/api/auth/organization/create`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': userAgent(),
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(input),
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(`Failed to create organization: ${res.status} ${text}`)
  }
  return (await res.json()) as { id: string }
}

export async function setActiveOrganization(
  baseUrl: string,
  token: string,
  organizationId: string,
): Promise<void> {
  const res = await fetch(`${baseUrl}/api/auth/organization/set-active`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'User-Agent': userAgent(),
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({ organizationId }),
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(`Failed to set active organization: ${res.status} ${text}`)
  }
}

File: chkit_python/src/chkit_plugin_obsessiondb/api_client.py
"""ObsessionDB auth API client (HTTP).

1:1 port of ``packages/plugin-obsessiondb/src/auth/api-client.ts``.

Endpoints covered:

- ``POST /api/auth/device/code`` — start device-code flow
- ``POST /api/auth/device/token`` — poll for the access token
- ``GET  /api/auth/get-session`` — read user + active org from a token
- ``POST /api/auth/email-otp/send-verification-otp`` — passwordless step 1
- ``POST /api/auth/sign-in/email-otp`` — passwordless step 2 (returns
  the bearer in the ``set-auth-token`` response header)
- ``POST /api/auth/organization/create``
- ``POST /api/auth/organization/set-active``

The service / jobs / workbench oRPC contracts are not in this turn; they
land alongside the service commands.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any

import httpx
from pydantic import BaseModel, ConfigDict

CLIENT_ID = "chkit-cli"
USER_AGENT = "chkit-cli"
HTTP_TIMEOUT_SECONDS = 30.0
HTTP_429_RATE_LIMITED = 429


class SessionExpiredError(RuntimeError):
    def __init__(self) -> None:
        super().__init__(
            "Session expired. Run `chkit obsessiondb login` to re-authenticate."
        )


class OtpRateLimitError(RuntimeError):
    """Raised when the send-OTP endpoint returns HTTP 429."""

    def __init__(self) -> None:
        super().__init__(
            "Too many code requests. Please wait a minute and try again."
        )


def is_session_expired_error(error: BaseException) -> bool:
    return isinstance(error, SessionExpiredError)


@dataclass(frozen=True, slots=True)
class DeviceCodeResponse:
    device_code: str
    user_code: str
    verification_uri: str
    verification_uri_complete: str
    expires_in: int
    interval: int


class SessionUser(BaseModel):
    model_config = ConfigDict(frozen=True, extra="ignore")

    id: str
    name: str
    email: str


class SessionInfo(BaseModel):
    model_config = ConfigDict(frozen=True, extra="ignore", populate_by_name=True)

    active_organization_id: str | None = None


class SessionResponse(BaseModel):
    model_config = ConfigDict(frozen=True, extra="ignore")

    user: SessionUser
    session: SessionInfo | None = None


class VerifiedUser(BaseModel):
    model_config = ConfigDict(frozen=True, extra="ignore")

    id: str
    email: str
    name: str | None = None


@dataclass(frozen=True, slots=True)
class OtpVerifyResult:
    token: str
    user: VerifiedUser


def _default_headers(token: str | None = None) -> dict[str, str]:
    headers = {
        "Content-Type": "application/json",
        "User-Agent": USER_AGENT,
    }
    if token is not None:
        headers["Authorization"] = f"Bearer {token}"
    return headers


def _client() -> httpx.Client:
    return httpx.Client(timeout=HTTP_TIMEOUT_SECONDS)


def request_device_code(base_url: str) -> DeviceCodeResponse:
    """Start an RFC 8628 device-code flow. Returns the user code + verification URI."""
    with _client() as http:
        res = http.post(
            f"{base_url}/api/auth/device/code",
            headers=_default_headers(),
            json={"client_id": CLIENT_ID},
        )
    if res.status_code >= httpx.codes.BAD_REQUEST:
        msg = f"Failed to request device code: {res.status_code} {res.text}"
        raise RuntimeError(msg)
    body = res.json()
    return DeviceCodeResponse(
        device_code=str(body["device_code"]),
        user_code=str(body["user_code"]),
        verification_uri=str(body["verification_uri"]),
        verification_uri_complete=str(body["verification_uri_complete"]),
        expires_in=int(body["expires_in"]),
        interval=int(body["interval"]),
    )


def poll_device_token(
    base_url: str,
    device_code: str,
    interval: float,
    expires_in: float,
    *,
    sleep: Any = time.sleep,
    monotonic: Any = time.monotonic,
) -> str:
    """Poll until the user authorises the device. Returns the access token."""
    deadline = monotonic() + expires_in
    poll_interval = interval

    while monotonic() < deadline:
        sleep(poll_interval)
        with _client() as http:
            res = http.post(
                f"{base_url}/api/auth/device/token",
                headers=_default_headers(),
                json={
                    "client_id": CLIENT_ID,
                    "device_code": device_code,
                    "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
                },
            )
        try:
            body = res.json()
        except ValueError:
            body = {}
        access_token = body.get("access_token")
        error = body.get("error")
        if not access_token and not error:
            msg = f"Token poll failed: {res.status_code} {body}"
            raise RuntimeError(msg)
        if access_token:
            return str(access_token)
        if error == "authorization_pending":
            continue
        if error == "slow_down":
            poll_interval += 5
            continue
        if error == "access_denied":
            msg = "Authorization denied by user."
            raise RuntimeError(msg)
        if error == "expired_token":
            msg = "Device code expired. Please try again."
            raise RuntimeError(msg)
        msg = f"Unexpected token poll response: {body}"
        raise RuntimeError(msg)

    msg = "Device code expired. Please try again."
    raise RuntimeError(msg)


def get_session(base_url: str, token: str) -> SessionResponse:
    """Read ``/api/auth/get-session`` → user + active organisation."""
    with _client() as http:
        res = http.get(
            f"{base_url}/api/auth/get-session",
            headers=_default_headers(token),
        )
    if res.status_code >= httpx.codes.BAD_REQUEST:
        msg = f"Failed to get session: {res.status_code} {res.text}"
        raise RuntimeError(msg)
    return SessionResponse.model_validate(res.json())


def send_verification_otp(base_url: str, email: str) -> None:
    """Trigger the passwordless OTP email. ``type='sign-in'`` covers signup + login."""
    with _client() as http:
        res = http.post(
            f"{base_url}/api/auth/email-otp/send-verification-otp",
            headers=_default_headers(),
            json={"email": email, "type": "sign-in"},
        )
    if res.status_code == HTTP_429_RATE_LIMITED:
        raise OtpRateLimitError
    if res.status_code >= httpx.codes.BAD_REQUEST:
        msg = f"Failed to send verification code: {res.status_code} {res.text}"
        raise RuntimeError(msg)


def verify_otp(base_url: str, email: str, otp: str) -> OtpVerifyResult:
    """Verify the OTP. The bearer token is in the ``set-auth-token`` response header."""
    with _client() as http:
        res = http.post(
            f"{base_url}/api/auth/sign-in/email-otp",
            headers=_default_headers(),
            json={"email": email, "otp": otp},
        )
    if res.status_code >= httpx.codes.BAD_REQUEST:
        msg = f"Failed to verify code: {res.status_code} {res.text}"
        raise RuntimeError(msg)
    token = res.headers.get("set-auth-token")
    if not token:
        msg = "Verification succeeded but no auth token was returned by the server."
        raise RuntimeError(msg)
    body = res.json()
    user_payload = body.get("user", {})
    return OtpVerifyResult(token=token, user=VerifiedUser.model_validate(user_payload))


def create_organization(
    base_url: str, token: str, *, name: str, slug: str
) -> dict[str, Any]:
    with _client() as http:
        res = http.post(
            f"{base_url}/api/auth/organization/create",
            headers=_default_headers(token),
            json={"name": name, "slug": slug},
        )
    if res.status_code >= httpx.codes.BAD_REQUEST:
        msg = f"Failed to create organization: {res.status_code} {res.text}"
        raise RuntimeError(msg)
    body: Any = res.json()
    if not isinstance(body, dict):
        return {}
    return body


def set_active_organization(
    base_url: str, token: str, organization_id: str
) -> None:
    with _client() as http:
        res = http.post(
            f"{base_url}/api/auth/organization/set-active",
            headers=_default_headers(token),
            json={"organizationId": organization_id},
        )
    if res.status_code >= httpx.codes.BAD_REQUEST:
        msg = f"Failed to set active organization: {res.status_code} {res.text}"
        raise RuntimeError(msg)
